source/test/TestApp/PlayFabQoSTest.cpp

```diff
@@ -24,11 +25,29 @@ namespace PlayFabUnit
 {
 #if (!UNITY_IOS && !UNITY_ANDROID) && (!defined(PLAYFAB_PLATFORM_IOS) && !defined(PLAYFAB_PLATFORM_ANDROID) && !defined(PLAYFAB_PLATFORM_SWITCH))
     /// QoS API
-    void PlayFabQoSTest::QosResultApi(TestContext& testContext)
+    void PlayFabQoSTest::QoSResultApi(TestContext& testContext)
     {
+        LoginWithCustomIDRequest request;
+        request.CustomId = PlayFabSettings::buildIdentifier;
+        request.CreateAccount = true;
+
+        std::atomic_bool isLoggedIn = false;
+
+        PlayFabClientAPI::LoginWithCustomID(request,
+            [&isLoggedIn](const LoginResult& loginResult, void*) {
+                PlayFabSettings::staticPlayer->clientSessionTicket = loginResult.SessionTicket;
+                isLoggedIn.exchange(true);
+            },
+            [&isLoggedIn](const PlayFabError&, void*) {
+                isLoggedIn.exchange(true);
+            },
+                &testContext);
+
+        while (!isLoggedIn) {}
```

Comment:
It's not appropriate to block infinitely on login.
The http layer, or the login, or whatever, could crash somewhere, and never set this boolean.

The C++ test framework is designed to do async testing where this method kicks off the test, and then the test comes back later with a result.
Review ClientApiTests, and follow that pattern.

Reply:
+1 and also you need to yield or sleep to give a chance to another thread to do its work instead of busy waiting.

---
In reply to: [514816352](https://github.com/PlayFab/XPlatCoreTemplate/pull/105#discussion_r514816352) [](ancestors = 514816352)

Reply:
The initial test was supposed to work in the same way as ClientApiTests, but due to the implementation of QoS, the multiplayer API call getting queued in a login call back leads to a dead lock. 

How? The thread that calls back on login tells qos to make that multiplayer call (for this tests purpose). But that is a critical point of failure for QoS since inception. QoS calls a multiplayer API to get back responses from servers. That multiplayer call gets queued on the WinHttp object's pending_requests. The issue here is, we are already on that WinHttp worker thread, completing the login call. 

Qos has always done this "block the current thread while we get callbacks to multiple playfab multiplayer calls and sort em". This API was not implemented appropriately for async, but that complete change is out of scope of this PR. 

This PR just covers getting the test passing to confirm behavior (it never passed before due to the deadlock, which probably coincides with changing over to the Instance based api's). QoS relies on the static api logic right now and that change is VERY out of scope to the particular bug we are looking at (if things time out, we should hand back a time out error code). 

We can address fixing this infinite blocking issue by re-architechting QoS to enable an instance based API version that accepts an authentication context that is required to make the multiplayer calls. That should remove the oddity of needing to wait anywhere (the qos instance should handle the callbacks for the multiplayer call, and then forward the final callback back to the customer like all of our other api's). 

---
In reply to: [514816352](https://github.com/PlayFab/XPlatCoreTemplate/pull/105#discussion_r514816352) [](ancestors = 514816352)